### DISCUSSION TITLE
Start Linux, then execute some scripts.

### DISCUSSION THREAD

**Discussion by guettli** [ipxe/ipxe#1079] [General]
I want to start a Linux in RAM via iPXE and then a script (loaded via http) should get executed.

iPXE works fine. Example:

```
#!ipxe

# Unfortunately "https" will fail with current iPXE with "operation not permitted"
# https://ipxe.org/err/410de1
set base-url http://dl-cdn.alpinelinux.org/alpine/v3.18/releases/x86_64/netboot-3.18.4

kernel ${base-url}/vmlinuz-virt console=tty0 modules=loop,squashfs quiet nomodeset alpine_repo=https://dl-cdn.alpinelinux.org/alpine/v3.18/main modloop=${base-url}/modloop-virt
initrd ${base-url}/initramfs-virt
boot
```

I tried to provide an alpine overlay via apkovl kernel parameter, but this fails:

https://serverfault.com/questions/1147870/alpine-boot-via-ipxe-and-execute-a-script-cant-stat-filesystem-no-such-fil

I am not married with Alpine. It could be any other small Linux distro.

I could create a custom initrd, like explained [here](https://unix.stackexchange.com/questions/689678/automate-alpine-linux-installation), but maybe there is an easier solution?

My preferred solution would look like this:

```
kernel .... init="wget -O- https://example.com/myscript.sh | sh"
```

But this does not work, since `init` needs to be binary, not a command.

I don't need many tools for the script: A shell, fsdisk, mkfs.ext4, chroot.

Do you have some hints for me how to start a simple Linux in RAM?

**Comment by NiKiZe** [comment]
In part duplicate of #1078 
iPXE can build combined initrds for you, so you can add script o iPXE, just have to figure out how to execute it?

**Comment by mcb30** [comment]
See https://ipxe.org/cmd/imgfetch?redirect=1#notes

iPXE can insert arbitrary downloaded files into any existing directory within the initrd image (as shown on that page).  Most distros will have some feature that executes a script of a specific name (or possibly execute all scripts found within a specific directory).  For example, TinyCore linux will execute `/opt/bootlocal.sh` if that file exists.  It's very likely that Alpine will do something similar.

**Comment by guettli** [comment]
@mcb30 thank you for your reply. I  have question about the example from example from the docs:

```
  #!ipxe
  
  kernel vmlinuz
  initrd initrd.img
  initrd initscripts/deploy.sh /sbin/init mode=755
  boot
```

Where is the source of `initscripts/deploy.sh`? 

Is it part of `initrd.img`?

**Reply by mcb30** [reply]
No, in that example `initscripts/deploy.sh` is a shell script downloaded from the same server as the iPXE script itself (since the URL `initscripts/deploy.sh` is a relative URL), and made to appear as `/sbin/init` within the initrd as seen by the kernel.

The kernel sees a virtual CPIO archive constructed on the fly by iPXE, which is functionally equivalent to what you would get by unpacking `initrd.img`, copying in `deploy.sh` as `/sbin/init`, and repacking it.

**Reply by guettli** [reply]
If I download the ipxe script from http://example.com/foo/file.ipxe, could I add an extra file like this?

```
  #!ipxe
  
  kernel vmlinuz
  initrd initrd.img
  initrd http://other-domain.com/bar/deploy.sh /sbin/init mode=755
  boot
```

**Reply by mcb30** [reply]
Yes, you can do that. There's nothing that requires you to download all files from the same server.
